### PR TITLE
add PAG support sd15 controlnet

### PR DESCRIPTION
**Pull request by tuanh123789**
# What does this PR do?

Adds PAG (Perturbed-Attention Guidance) support for SD 1.5 with controlnet models (StableDiffusionControlNetPAGPipeline)
Continuation of #8710

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@yiyixuxu 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

##
| Sample using SD1.5 controlnet | Sample SD1.5 controlnet PAG |
|-------------------------------|----------------------------|
| test | test_10 |

```
from diffusers import AutoPipelineForText2Image, ControlNetModel, UniPCMultistepScheduler
from diffusers.utils import load_image
import numpy as np
import torch

import cv2
from PIL import Image

# download an image
image = load_image(
   "https://hf.co/datasets/hf-internal-testing/diffusers-images/resolve/main/sd_controlnet/hf-logo.png"
)
image = np.array(image)

# get canny image
image = cv2.Canny(image, 100, 200)
image = image[:, :, None]
image = np.concatenate([image, image, image], axis=2)
canny_image = Image.fromarray(image)

# load control net and stable diffusion v1-5
controlnet = ControlNetModel.from_pretrained("lllyasviel/sd-controlnet-canny", torch_dtype=torch.float16)
pipe = AutoPipelineForText2Image.from_pretrained(
   "runwayml/stable-diffusion-v1-5", controlnet=controlnet, torch_dtype=torch.float16, enable_pag=True
)

# speed up diffusion process with faster scheduler and memory optimization
# remove following line if xformers is not installed
pipe.enable_xformers_memory_efficient_attention()

pipe.enable_model_cpu_offload()

# generate image
generator = torch.manual_seed(0)
image = pipe(
   "aerial view, a futuristic research complex in a bright foggy jungle, hard lighting",
   guidance_scale=7.5,
   generator=generator,
   image=canny_image,
   pag_scale=10,
).images[0]
```

### CODE REVIEW

File: src/diffusers/pipelines/__init__.py [hunk 1]
```diff
@@ -248,6 +249,7 @@
         "StableDiffusion3Img2ImgPipeline",
         "StableDiffusion3InpaintPipeline",
     ]
+    _import_structure["stable_diffusion_3"] = ["StableDiffusion3Pipeline", "StableDiffusion3Img2ImgPipeline"]
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/__init__.py:252]:
```suggestion
```

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py [hunk 2]
```diff
@@ -0,0 +1,1343 @@
...
+                        negative_image_embeds, image_embeds = image_embeds.chunk(2)
+
+                    if self.do_perturbed_attention_guidance:
+                        image_embeds = self._prepare_perturbed_attention_guidance(
+                            image_embeds, negative_image_embeds, self.do_classifier_free_guidance
+                        )
+                    elif self.do_classifier_free_guidance:
+                        image_embeds = torch.cat([negative_image_embeds, image_embeds], dim=0)
+                    image_embeds = image_embeds.to(device)
+                    ip_adapter_image_embeds[i] = image_embeds
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py:1188]: ip_adapter_image_embeds will not be None after prepare

```suggestion
            for i, image_embeds in enumerate(ip_adapter_image_embeds):
                negative_image_embeds = None
                if self.do_classifier_free_guidance:
                    negative_image_embeds, image_embeds = image_embeds.chunk(2)

                if self.do_perturbed_attention_guidance:
                    image_embeds = self._prepare_perturbed_attention_guidance(
                        image_embeds, negative_image_embeds, self.do_classifier_free_guidance
                    )
                elif self.do_classifier_free_guidance:
                    image_embeds = torch.cat([negative_image_embeds, image_embeds], dim=0)
                image_embeds = image_embeds.to(device)
                ip_adapter_image_embeds[i] = image_embeds
```

**Review by a-r-r-o-w — COMMENTED** [review]
Looking great. Just small changes remaining and this should be good for merge.

File: src/diffusers/pipelines/__init__.py (hunk 1 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/__init__.py:252]: The failing tests are also caused due to this

**Comment by a-r-r-o-w** [comment]
Thank you for addressing the comments! Could we fix the failing tests?

<details>
<summary> Error logs </summary>

```
...
FAILED tests/pipelines/pag/test_pag_controlnet_sd.py::StableDiffusionControlNetPAGPipelineFastTests::test_pag_cfg - AssertionError: output is different from expected, [0.45505235 0.2785938  0.16334778 0.79689944 0.53095645 0.40135607
   0.7052706  0.69065094 0.41548574]
assert 0.3865616192955017 < 0.001
FAILED tests/pipelines/pag/test_pag_controlnet_sd.py::StableDiffusionControlNetPAGPipelineFastTests::test_pag_uncond - AssertionError: output is different from expected, [0.45127502 0.2797252  0.15970308 0.7993157  0.5414344  0.40160775
   0.7114598  0.69803864 0.4217583 ]
assert 0.39445382411422725 < 0.001
FAILED tests/pipelines/pag/test_pag_controlnet_sd.py::StableDiffusionControlNetPAGPipelineFastTests::test_save_load_optional_components - KeyError: 'tokenizer_2'
```
</details>

Let me know if you need any help. We can merge as soon as these are fixed since everything else looks good to me!

**Comment by tuanh123789** [comment]
Sure can you help me on tests 🤗

**Comment by a-r-r-o-w** [comment]
> Sure can you help me on tests 🤗

For the first two test failures, you will have to update the expected values to what is the actual value of `image_slice` is. You can print the values of image_slice and replace expected_slice with that :stuck_out_tongue_winking_eye: 

For the third test, you need to remove `SDXLOptionalComponentsTesterMixin` in the fast tests class.

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py [hunk 3]
```diff
@@ -0,0 +1,1342 @@
...
+    def decode_latents(self, latents):
+        deprecation_message = "The decode_latents method is deprecated and will be removed in 1.0.0. Please use VaeImageProcessor.postprocess(...) instead"
+        deprecate("decode_latents", "1.0.0", deprecation_message, standard_warn=False)
+
+        latents = 1 / self.vae.config.scaling_factor * latents
+        image = self.vae.decode(latents, return_dict=False)[0]
+        image = (image / 2 + 0.5).clamp(0, 1)
+        # we always cast to float32 as this does not cause significant overhead and is compatible with bfloat16
+        image = image.cpu().permute(0, 2, 3, 1).float().numpy()
+        return image
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py:544]:
```suggestion
```
I don't think it is used

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py [hunk 4]
```diff
@@ -0,0 +1,1342 @@
...
+        if accepts_eta:
+            extra_step_kwargs["eta"] = eta
+
+        # check if the scheduler accepts generator
+        accepts_generator = "generator" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py:564]: should we add a `#Copied from` here

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Comment by yiyixuxu** [comment]
feel free to merge once you're happy with it! @a-r-r-o-w

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py (hunk 4 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py:564]: I don't think we have a version of check_inputs with `callback_steps` removed, and this is the first of its kind

**Comment by tuanh123789** [comment]
@a-r-r-o-w Can you help me with last fail of test  🤗

**Comment by a-r-r-o-w** [comment]
@tuanh123789 The failing test is unrelated. I'm happy to merge this once you address YiYi's comment: https://github.com/huggingface/diffusers/pull/8820#discussion_r1674676589

**Comment by tuanh123789** [comment]
> @tuanh123789 The failing test is unrelated and nothing needs to be done on your end. I'm happy to merge this once you address YiYi's comment: [#8820 (comment)](https://github.com/huggingface/diffusers/pull/8820#discussion_r1674676589)

:v Sure, will be done right away

**Review by a-r-r-o-w — APPROVED** [review]
Thanks, looking good! One other thing that I just realised - could you update the pipelines doc page with this SD15 Controlnet? https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/pag.md. Please see #7944 for an example

**Comment by tuanh123789** [comment]
> Thanks, looking good! One other thing that I just realised - could you update the pipelines doc page with this SD15 Controlnet? https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/pag.md. Please see #7944 for an example

Done

**Comment by crapthings** [comment]
pip install git+https://github.com/huggingface/diffusers

ImportError: cannot import name 'StableDiffusionControlNetPAGPipeline' from 'diffusers' 

pip list

diffusers                 0.30.0.dev0


how can i test?

**Comment by crapthings** [comment]
how do i know if its working

fix seed and test?

```
pip uninstall diffusers
pip install git+https://github.com/huggingface/diffusers
```

**Comment by tuanh123789** [comment]
from diffusers import AutoPipelineForText2Image, ControlNetModel, UniPCMultistepScheduler
from diffusers.utils import load_image
import numpy as np
import torch

import cv2
from PIL import Image

# download an image
image = load_image(
   "https://hf.co/datasets/hf-internal-testing/diffusers-images/resolve/main/sd_controlnet/hf-logo.png"
)
image = np.array(image)

# get canny image
image = cv2.Canny(image, 100, 200)
image = image[:, :, None]
image = np.concatenate([image, image, image], axis=2)
canny_image = Image.fromarray(image)

# load control net and stable diffusion v1-5
controlnet = ControlNetModel.from_pretrained("lllyasviel/sd-controlnet-canny", torch_dtype=torch.float16)
pipe = AutoPipelineForText2Image.from_pretrained(
   "runwayml/stable-diffusion-v1-5", controlnet=controlnet, torch_dtype=torch.float16, enable_pag=True
)

# speed up diffusion process with faster scheduler and memory optimization
# remove following line if xformers is not installed
pipe.enable_xformers_memory_efficient_attention()

pipe.enable_model_cpu_offload()

# generate image
generator = torch.manual_seed(0)
image = pipe(
   "aerial view, a futuristic research complex in a bright foggy jungle, hard lighting",
   guidance_scale=7.5,
   generator=generator,
   image=canny_image,
   pag_scale=10,
).images[0]

**Comment by Bhavay-2001** [comment]
Hi @tuanh123789, could you please share the colab notebook you used to test the images which you have shown in the above output. Thanks

**Comment by asomoza** [comment]
> fix seed and test?

yes, preferably with a complex generation that you can see pag helps to fix.

> could you please share the colab notebook you used to test the images which you have shown in the above output

It's literally copy & paste the example from this PR, there's nothing else you have to do. 

|without pag|with pag|
|---|---|
|result_normal|result_pag|

You could just paste everything in one cell block, but here's  [notebook](https://colab.research.google.com/drive/1k8cYN3vxwliiHho-mrw6JbM_N1WeW4Q5?usp=sharing) with the steps to test it.

**Pull request merged**